Expand documentation for now APE14 WCS API

astrofrog opened the pull request:
This adds some basic usage docs for the new WCS APE 14 API

One important addition - it implements ``WCS.array_shape`` which means that we'll finally be able to close e.g. https://github.com/astropy/astropy/issues/5461 or https://github.com/astropy/astropy/issues/4669 - at least I think.

Hey @drdavella - I really could do with https://github.com/astropy/pytest-doctestplus/issues/2 here :wink: (not critical for the PR as I've just put ``+REMOTE_DATA`` everywhere, just pointing out a good use case!)

Once the CI runs, there will be a link to preview the docs (from 'giles') in the status checks below.

[Direct link to docs preview](https://13279-2081289-gh.circle-artifacts.com/0/home/circleci/project/docs/_build/html/wcs/wcsapi.html)

eteq commented:
One quick thing not directly related to the review: @nden - can you confirm the `. array_shape` as implemented here is compatible with/could be made so with `gwcs` (due to e.g. https://github.com/astropy/astropy/issues/5461#issuecomment-307796773)?

astrofrog commented on astropy/wcs/wcsapi/fitswcs.py:
```diff
@@ -121,6 +121,13 @@ def pixel_n_dim(self):
         return len(self.wcs.ctype)
 
+    @property
+    def array_shape(self):
```
It's in array/numpy order - all these properties have docstrings in the base API that get inherited, that's why there is apparently no docstring here (but users would see one). See https://github.com/astropy/astropy/blob/master/astropy/wcs/wcsapi/low_level_api.py#L196

In these docs, I've also tried to clarify that anything with 'array' in the name is in array index order.

astrofrog replied:
Note that we could add ``pixel_shape`` if you wanted

nden replied:
:+1: for adding `pixel_shape` because its order corresponds to the `NAXIS*` order. 

I think we should deprecate the setters for the private `_naxis1/2` attributes and change `calc_footprint` to use `pixel_shape`. I can do this in a separate PR if this sounds acceptable.

eteq replied:
I'm also 👍 to adding `pixel_shape`.  While it seems a bit picky it seems consistent with the other methods now.

astrofrog replied:
Added

nden commented on docs/wcs/wcsapi.rst:
```diff
@@ -22,7 +22,128 @@ in Python. This API is described in the Astropy Proposal for Enhancements (APE)
+the method or property used, and this can normally be determined from the
+property or method name. Properties and methods containing ``pixel`` assume
+``(x, y)`` ordering, while properties and methods containing ``array`` assume
+``(row, column)`` ordering.
```
Would it help to say:
```
Properties and methods containing ``pixel`` assume
``(x, y)`` or ``Carstesian`` ordering, while properties and methods containing ``array`` assume
``(row, column)`` or ``numpy array`` ordering.

```

eteq replied:
I guess that doesn't help because as our discussions revealed, I don't really get what "Cartesian ordering" means: to me "Cartesian" means "the metric is s^2 = x^2 + y^2", which is true regardless of the ordering.  And "numpy array" also doesn't help because it can be both fortran *or* C ordered.

If preferred I can write the words on it... once I actually have right in my head which is supposed to be which :wink:

astrofrog replied:
I'll push some changes soon that are similar to @nden's suggestion.

astrofrog commented on docs/wcs/wcsapi.rst:
```diff
@@ -22,7 +22,128 @@ in Python. This API is described in the Astropy Proposal for Enhancements (APE)
+===========
+
+Let's take a look at the shared Python interface for WCS by using a spectral
+cube (two celestial axes and one spectral axis)::
```
Done

astrofrog commented:
@eteq - I've expanded the pixel convention section. Does it make sense now? I also added a basic image example.

@nden - I added ``pixel_shape`` for consistency. Please feel free to do another PR to deprecate the naxis1/naxis2 access as you mentioned.

astrofrog commented:
@eteq - I think this is ready for a final review/merge

nden commented:
@eteq In gwcs I have a setter for `array_shape` because the interface to `gwcs.WCS` does not currently take a header like `astropy.wcs.WCS`. In the future, when there's a more user friendly interface this will change but I imagine the implementation will give the same results. The comment you linked to was referring to the equivalent of `pixel_bounds`. Now that I look at it, `astropy.wcs.WCS.pixel_bounds`is a property which returns `None` but is not settable.  In gwcs `pixel_bounds` returns the `bounding_box` of the model. There's no equivalent attribute in `astropy.wcs`. So perhaps this property should have a setter in astropy?

nden commented:
Here's a potential issue with not having setters for `array_shape` and `pixel_shape`. 
There are setters for the private `_naxis1` and `_naxis2` attributes which we want to replace with the public `pixel_shape`. I can think of at least one use case where being able to set `pixel_shape` is useful. When drizzling images a common output WCS is created based on the WCSs of the input images. This output WCS needs to be assigned a `pixel_shape` value, normally computed using `pixel_shape`  of all input WCSs.

astrofrog commented:
@nden - we can easily add setters on the FITS WCS mix-in, but I think it would make more sense for you to do it in a separate PR as you know how best to set ``_naxis``

nden commented:
@eteq I added `pixel_bounds` setter.

astrofrog commented:
Ok so at this point this just needs a suggestion from @eteq on improving wording about the pixel convention stuff - but this can be merged without that change if it comes to it, as that can be clarified after feature freeze.

eteq commented:
Yeah, I agree with @astrofrog that it's better to merge this and we can have a follow-on clarifying PR if/when we get to it.  Presumably this wording works for *some* people. (At the very least, people that think like @nden and @astrofrog, so it's definitely better than nothing!)
